tka/deeplink.go

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package tka
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"net/url"
+	"strings"
+)
+
+const (
+	DeeplinkTailscaleURLScheme = "tailscale"
+	DeeplinkCommandSign        = "sign-device"
+)
+
+type DeeplinkValidationResult struct {
+	IsValid      bool
+	Error        error
+	Version      uint8
+	NodeKey      string
+	TLPub        string
+	DeviceName   string
+	OSName       string
+	EmailAddress string
+}
+
+// GenerateHMAC computes a SHA-256 HMAC for the concatenation of components, using
+// stateID as secret. It returns the result as a hex-encoded string.
+func generateHMAC(stateID uint64, components []string) string {
+	key := make([]byte, 8)
+	binary.LittleEndian.PutUint64(key, stateID)
+	mac := hmac.New(sha256.New, key)
+	for _, component := range components {
+		mac.Write([]byte(component))
+	}
+	return hex.EncodeToString(mac.Sum(nil))
+}
+
+// ValidateDeeplink validates a device signing deeplink using the given authority stateID.
+// The input urlString follows this structure:
+//
+// tailscale://sign-device/v1/?nk=xxx&tp=xxx&dn=xxx&os=xxx&em=xxx&hm=xxx
+//
+// where:
+// - "nk" is the nodekey of the node being signed
+// - "tp" is the tailnet lock public key
+// - "dn" is the name of the node
+// - "os" is the operating system of the node
+// - "em" is the email address associated with the node
+// - "hm" is a SHA-256 HMAC computed over the concatenation of the above fields
+func ValidateDeeplink(urlString string, stateID uint64) DeeplinkValidationResult {
```

Comment:
Maybe make this a method on a `*tka.Authority`, so it can just get the stateID automatically?

Reply:
done